parlai/agents/vsepp_caption/modules.py

```diff
@@ -0,0 +1,215 @@
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+import torch
+from torch import optim
+import torch.nn as nn
+import torchvision.models as models
+
+from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence
+
+import numpy as np
+
+
+class VSEpp(nn.Module):
+    """
+    Implementation of Visual Semantic Embedding++ model borrowed heavily from
+    here: <https://github.com/fartashf/vsepp>
+    """
+
+    def __init__(self, opt, dict, use_cuda):
+        super().__init__()
+        self.opt = opt
+        self.dict = dict
+        self.img_enc = EncoderImage(embed_size=opt['embed_size'],
+                                    finetune=opt['finetune'],
+                                    cnn_type=opt['cnn_type'],
+                                    no_imgnorm=opt['no_imgnorm'])
+        self.txt_enc = EncoderText(vocab_size=len(self.dict.tok2ind),
+                                   word_dim=opt['word_dim'],
+                                   embed_size=opt['embed_size'],
+                                   num_layers=opt['num_layers'],
+                                   use_cuda=use_cuda)
+        if use_cuda:
+            self.img_enc.cuda()
+            self.txt_enc.cuda()
+
+    def forward(self, images, captions, lengths):
+        img_emb = self.img_enc(images) if images is not None else None
+        cap_emb = self.txt_enc(captions, lengths) if captions is not None else None
+        return img_emb, cap_emb
```

Comment:
you can also do `images and self.img_enc(images)` for the same effect as the ternary expressions if you like the flavor better

Reply:
Didn't know about that!